README
$Id: README 1885379 2021-01-12 09:32:59Z ssteiner $

==============================================================================
  APACHE XML GRAPHICS COMMONS  -  README
==============================================================================

Contents of this file:
 - What is Apache XML Graphics Commons?
 - Where to get help?
 - How do I build Apache XML Graphics Commons?
 - Legal information
 - Release Notes

==============================================================================


What is Apache XML Graphics Commons?
---------------------------------------

It is a place where the Apache Batik and Apache FOP share commonly used 
components. Many components in here are also usable and useful stand-alone.
    
Documentation
---------------------    

The documentation can be found under:
http://xmlgraphics.apache.org/commons/

Where to get help?
---------------------    

Please subscribe to the general@xmlgraphics.apache.org mailing list by sending
an empty mail to general-subscribe@xmlgraphics.apache.org. All you questions
about XML Graphics Commons will be answered there.    

Please report bugs to JIRA at http://issues.apache.org/jira/ (project
XMLGraphicsCommons).


How do I build Apache XML Graphics Commons?
----------------------------------------------

If you've downloaded a binary distribution of Apache FOP or Apache Batik, 
you don't need to build XML Graphics Commons. It is distributed with these
two products.

Otherwise, install Apache Ant (http://ant.apache.org) and run "ant" from the
command-line in the directory where you find "build.xml".

Legal information
--------------------

Apache XML Graphics Commons is published under the Apache License 
version 2.0. For the license text, please see the following files:
- LICENSE
- NOTICE

Legal information on libraries used by Apache XML Graphics Commons, i.e. 
its dependencies, can be found in the "lib/README.txt" file.

Apache XML Graphics Commons contains unmodified copies of the Adobe Glyph
List and the ITC Zapf Dingbats Glyph List, available from:
http://www.adobe.com/devnet/opentype/archives/glyph.html

License labeling (according to http://www.apache.org/legal/3party.html):
- Multi-Licensed
- Source Available
- No Reciprocity Required

==============================================================================
  KNOWN ISSUES
==============================================================================

- PostScript output: When TexturePaint is used with PSGraphics2D, the output
  violates the DSC specification. Extracting single pages from a multi-page
  PostScript file can result in errors.

==============================================================================
  RELEASE NOTES
==============================================================================

For more detailed info about the changes, please see:
http://xmlgraphics.apache.org/commons/changes.html

Version 2.6
--------------

Minor release to be used with FOP 2.6

Version 2.4
--------------

Minor release to be used with FOP 2.4

Version 2.3
--------------

Minor release to be used with FOP 2.3

Version 2.2
--------------

* Transition from Ant to Maven Build Process

Minor release to be used with FOP 2.2

Version 2.1
--------------

Minor release to be used with FOP 2.1

Version 2.0.1
--------------

Minor release to be used with FOP 2.0

Version 2.0
--------------

The main new feature is the introduction of a URI resolution framework that
makes it easier to control resource access in a cloud environment. Because of
changes to the API, the version number has been bumped to 2.0.

This release also contains a number of bug fixes.

Version 1.5
--------------

This release of Apache XML Graphics Commons  primarily addresses bug fixes
and also adds a number of new features.

Version 1.4
-----------

This release adds the option to generate smaller PostScript files,
support for the AdobeStandardCyrillic encoding, RefinedImageFlavor,
TexturePaint support for PSGraphics2D (PostScript tiling patterns),
improvements to the XMP framework, optimization for PostScript state
handling in (E)PSDocumentGraphics2D, and more. In addition it contains
a number of bug fixes.

Version 1.3.1
--------------

This release is mostly a bugfix release for the image loading framework that
has been introduced in version 1.3.

Version 1.3
--------------

The most important addition in this release is an image loading framework
which supports all sorts of different image formats (bitmap and vector) and
is highly extensible.
Besides that there were a larger number of smaller additions and bugfixes.
Support for Java 1.3 has been dropped. Java 1.4 or later is required now.

Version 1.2
--------------

This release mainly adds support for CMYK and GRAY color spaces for PSGenerator.

Version 1.1
--------------

This release adds an XMP metadata framework and brings improvements for the
ImageWriter package plus some minor fixes mainly in the PostScript area.

Version 1.0
--------------

This is the first release of Apache XML Graphics Commons. There are currently
no known issues with the code.


debian/changelog
xmlgraphics-commons (2.6-1) unstable; urgency=medium

  * New upstream version 2.6
  * d/patches: Remove patch applied upstream

 -- Mathieu Malaterre <malat@debian.org>  Wed, 17 Nov 2021 09:21:28 +0100

xmlgraphics-commons (2.4-3) unstable; urgency=medium

  * Team upload
  * Update Vcs URLs to point to Salsa
  * Use debhelper-compat 13
  * Remove deprecated get-orig-source target
  * Bump Standards-Version to 4.6.0
  * Set timezone to UTC when SOURCE_DATE_EPOCH is set (Closes: #983588)
    Thank you to Vagrant Cascadian for the patch.
  * Set Rules-Requires-Root: no in debian/control

 -- tony mancill <tmancill@debian.org>  Sun, 10 Oct 2021 14:49:19 -0700

xmlgraphics-commons (2.4-2) unstable; urgency=high

  * Team upload.
  * Fix CVE-2020-11988:
    Apache XmlGraphics Commons is vulnerable to server-side request forgery,
    caused by improper input validation by the XMPParser. By using a
    specially-crafted argument, an attacker could exploit this vulnerability to
    cause the underlying server to make arbitrary GET requests.
    (Closes: #984949)

 -- Markus Koschany <apo@debian.org>  Mon, 02 Aug 2021 07:48:42 +0200

xmlgraphics-commons (2.4-1) unstable; urgency=medium

  * New upstream version 2.4
  * Update Std-Vers to 4.5.0 no changes needed

 -- Mathieu Malaterre <malat@debian.org>  Fri, 14 Feb 2020 22:20:36 +0100

xmlgraphics-commons (2.3-1) unstable; urgency=medium

  * Team upload.
  * New upstream version 2.3.
    - Apparently all tests run successfully again. (Closes: #906523)
  * Switch to compat level 11.
  * Declare compliance with Debian Policy 4.2.0.
  * Remove README.source. Obsolete.
  * Remove Vincent Fourmond from Uploaders at this own request.
    (Closes: #897314)
  * Remove all patches because they were disabled and not used at all.

 -- Markus Koschany <apo@debian.org>  Wed, 22 Aug 2018 23:30:39 +0200

xmlgraphics-commons (2.2-1) unstable; urgency=medium

  * Team upload.
  * New upstream release (Closes: #858815)
  * Build with Maven instead of Ant
  * Moved the package to Git
  * Standards-Version updated to 4.1.1
  * Switch to debhelper level 10

 -- Emmanuel Bourg <ebourg@apache.org>  Tue, 03 Oct 2017 01:23:23 +0200

xmlgraphics-commons (2.1-2) unstable; urgency=medium

  * Team upload.
  * Add gnupg to Build-Depends. (Closes: #834744)
  * Declare compliance with Debian Policy 3.9.8.
  * Vcs-Browser: Use https.

 -- Markus Koschany <apo@debian.org>  Thu, 08 Sep 2016 04:31:18 +0200

xmlgraphics-commons (2.1-1) unstable; urgency=medium

  * New upstream release
  * Add PGP key check

 -- Mathieu Malaterre <malat@debian.org>  Mon, 01 Feb 2016 14:12:43 +0100

xmlgraphics-commons (2.0.1-2) unstable; urgency=medium

  * Upload to sid

 -- Mathieu Malaterre <malat@debian.org>  Sun, 13 Sep 2015 11:41:41 +0200

xmlgraphics-commons (2.0.1-1) experimental; urgency=medium

  * New upstream. Closes: #792131
  * Fix lintian error for -doc package.
  * Bump Std-Vers to 3.9.6 no changes needed

 -- Mathieu Malaterre <malat@debian.org>  Sat, 11 Jul 2015 22:12:01 +0200

xmlgraphics-commons (1.5-4) unstable; urgency=low

  [ Emmanuel Bourg ]
  * Team upload.
  * debian/control:
    - Removed junit4 and libmockito-java from the dependencies
      of libxmlgraphics-commons-java
    - Updated Standards-Version to 3.9.5 (no changes)
    - Use canonical URLs for the Vcs-* fields
    - Fixed the duplicate-short-description lintian warning
  * Switch to debhelper level 9
  * debian/rules: Improved the clean target
  * Include the Javadoc in the documentation package
    instead of the raw xdoc files
  * Install the documentation in /usr/share/doc/libxmlgraphics-commons-java

  [ tony mancill ]
  * debian/rules:
    - unexport DISPLAY to prevent test failures on workstations

 -- Emmanuel Bourg <ebourg@apache.org>  Thu, 07 Nov 2013 23:11:13 +0100

xmlgraphics-commons (1.5-3) unstable; urgency=low

  * Fix lintian error (doc-base-file-references-missing-file)
  * Enable testsuite + disable test which is incompat with openjdk-7.
    Thanks to James Page. Closes: #708616

 -- Mathieu Malaterre <malat@debian.org>  Wed, 29 May 2013 14:34:25 +0200

xmlgraphics-commons (1.5-2) unstable; urgency=low

  * Upload to sid
  * Bump Std-Vers to 3.9.4, no changes needed
  * Use updated debian/watch (HowToHelpWithFixingWatchFiles)

 -- Mathieu Malaterre <malat@debian.org>  Wed, 15 May 2013 14:43:50 +0200

xmlgraphics-commons (1.5-1) experimental; urgency=low

  * Use my @d.o alias, remove DMUA flag (not required)
  * Update the xmlgraphics-commons Uploaders list. Closes: #654142
  * Bump Std-Vers to 3.9.3, no changes needed
  * Fix VCS urls
  * Use javahelper for repack and d/rules
  * Build documentation package
  * remove XMP patches:
    - xml-rdf-resource.patch
    - xml-top-level.patch

 -- Mathieu Malaterre <malat@debian.org>  Mon, 29 Oct 2012 08:59:08 +0100

xmlgraphics-commons (1.4.dfsg-4) unstable; urgency=low

  * Team upload.
  * d/control: Drop Depends on java2-runtime-headless since only programs
    need to depends on Java runtime (Debian Java Policy).
  * Bump Standards-Version to 3.9.2: no changes needed.
  * d/rules: Fix JAVA_HOME for multiarch OpenJDK.
  * d/copyright: Update to latest DEP-5 format.

 -- Damien Raude-Morvan <drazzib@debian.org>  Fri, 25 Nov 2011 08:39:18 +0100

xmlgraphics-commons (1.4.dfsg-3) unstable; urgency=low

  * Uploading to unstable, hoping we won't break too many things ;-)...

 -- Vincent Fourmond <fourmond@debian.org>  Fri, 11 Feb 2011 14:15:14 +0100

xmlgraphics-commons (1.4.dfsg-2) experimental; urgency=low

  * Applied patches by Brian M. Carlson <sandals@crustytoothpaste.net>
    to fix various problems with XMP (closes: #605940, #605941)

src/test/java/org/apache/xmlgraphics/ps/ImageEncodingHelperTestCase.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* $Id: ImageEncodingHelperTestCase.java 1732019 2016-02-24 05:01:10Z gadams $ */

package org.apache.xmlgraphics.ps;

import java.awt.Color;
import java.awt.Graphics2D;
import java.awt.geom.Rectangle2D;
import java.awt.image.BufferedImage;
import java.awt.image.DataBuffer;
import java.awt.image.DirectColorModel;
import java.awt.image.WritableRaster;
import java.io.IOException;
import java.util.Arrays;

import org.junit.Test;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;

import static org.junit.Assert.assertTrue;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Matchers.anyObject;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import org.apache.commons.io.output.ByteArrayOutputStream;

public class ImageEncodingHelperTestCase {

    private BufferedImage prepareImage(BufferedImage image) {
        Graphics2D ig = image.createGraphics();
        ig.scale(.5, .5);
        ig.setPaint(new Color(128, 0, 0));
        ig.fillRect(0, 0, 100, 50);
        ig.setPaint(Color.orange);
        ig.fillRect(100, 0, 100, 50);
        ig.setPaint(Color.yellow);
        ig.fillRect(0, 50, 100, 50);
        ig.setPaint(Color.red);
        ig.fillRect(100, 50, 100, 50);
        ig.setPaint(new Color(255, 127, 127));
        ig.fillRect(0, 100, 100, 50);
        ig.setPaint(Color.black);
        ig.draw(new Rectangle2D.Double(0.5, 0.5, 199, 149));
        ig.dispose();
        return image;
    }

    /**
     * Tests encodeRenderedImageWithDirectColorModeAsRGB(). Tests the optimised method against the
     * non-optimised method(encodeRenderedImageAsRGB) to ensure the BufferedImage produced are the
     * same.
     * @throws IOException if an I/O error occurs.
     */
    @Test
    public void testEncodeRenderedImageWithDirectColorModelAsRGB() throws IOException {
        BufferedImage image = new BufferedImage(100, 75, BufferedImage.TYPE_INT_ARGB);
        image = prepareImage(image);

        ByteArrayOutputStream optimized = new ByteArrayOutputStream();
        ImageEncodingHelper.encodeRenderedImageWithDirectColorModelAsRGB(image, optimized);

        ByteArrayOutputStream nonoptimized = new ByteArrayOutputStream();
        ImageEncodingHelper.encodeRenderedImageAsRGB(image, nonoptimized);

        assertTrue(Arrays.equals(nonoptimized.toByteArray(), optimized.toByteArray()));

    }

    /**
     * Tests a BGR versus RBG image. Debugging shows the BGR follows the optimizeWriteTo() (which
     * is intended). The bytes are compared with the RBG image, which happens to follow the
     * writeRGBTo().
     *
     * @throws IOException
     */
    @Test
    public void testRGBAndBGRImages() throws IOException {
        BufferedImage imageBGR = new BufferedImage(100, 75, BufferedImage.TYPE_3BYTE_BGR);
        imageBGR = prepareImage(imageBGR);
        BufferedImage imageRGB = new BufferedImage(100, 75, BufferedImage.TYPE_INT_BGR);
        imageRGB = prepareImage(imageRGB);

        ImageEncodingHelper imageEncodingHelperBGR = new ImageEncodingHelper(imageBGR, false);
        ImageEncodingHelper imageEncodingHelperRGB = new ImageEncodingHelper(imageRGB, false);

        ByteArrayOutputStream baosBGR = new ByteArrayOutputStream();
        imageEncodingHelperBGR.encode(baosBGR);

        ByteArrayOutputStream baosRGB = new ByteArrayOutputStream();
        imageEncodingHelperRGB.encode(baosRGB);

        assertTrue(Arrays.equals(baosBGR.toByteArray(), baosRGB.toByteArray()));
    }

    /**
     * Tests encodeRenderedImageWithDirectColorModeAsRGB(). Uses mocking to test the method
     * implementation.
     * @throws IOException if an I/O error occurs.
     */
    @Test
    public void testMockedEncodeRenderedImageWithDirectColorModelAsRGB() throws IOException {
        BufferedImage image = mock(BufferedImage.class);
        final int[] templateMasks = new int[] {0x00ff0000 /*R*/, 0x0000ff00 /*G*/,
                                                    0x000000ff /*B*/, 0xff000000 /*A*/};
        DirectColorModel dcm = new DirectColorModel(255, templateMasks[0], templateMasks[1],
                templateMasks[2], templateMasks[3]);

        WritableRaster raster = mock(WritableRaster.class);
        DataBuffer buffer = mock(DataBuffer.class);

        when(image.getColorModel()).thenReturn(dcm);
        when(image.getRaster()).thenReturn(raster);
        when(raster.getDataBuffer()).thenReturn(buffer);
        when(buffer.getDataType()).thenReturn(DataBuffer.TYPE_INT);
        when(image.getWidth()).thenReturn(3);
        when(image.getHeight()).thenReturn(3);
        final int expectedValue = 1 + 2 << 8 + 3 << 16;
        Answer<Object> ans = new Answer<Object>() {
            public Object answer(InvocationOnMock invocation) throws Throwable {
                Object[] args = invocation.getArguments();
                int[] data = (int[]) args[4];
                Arrays.fill(data, expectedValue);
                return null;
            }
        };
        when(raster.getDataElements(anyInt(), anyInt(), anyInt(), anyInt(), anyObject()))
                .thenAnswer(ans);

        ByteArrayOutputStream optimized = new ByteArrayOutputStream();
        ImageEncodingHelper.encodeRenderedImageWithDirectColorModelAsRGB(image, optimized);

        byte[] expectedByteArray = new byte[27];
        Arrays.fill(expectedByteArray, (byte) expectedValue);
        assertTrue(Arrays.equals(expectedByteArray, optimized.toByteArray()));
    }
}


src/main/java/org/apache/xmlgraphics/ps/ImageEncodingHelper.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* $Id: ImageEncodingHelper.java 1713619 2015-11-10 11:55:53Z ssteiner $ */

package org.apache.xmlgraphics.ps;

import java.awt.color.ColorSpace;
import java.awt.image.BufferedImage;
import java.awt.image.ColorModel;
import java.awt.image.ComponentColorModel;
import java.awt.image.DataBuffer;
import java.awt.image.DataBufferByte;
import java.awt.image.DirectColorModel;
import java.awt.image.IndexColorModel;
import java.awt.image.PixelInterleavedSampleModel;
import java.awt.image.Raster;
import java.awt.image.RenderedImage;
import java.awt.image.SampleModel;
import java.io.IOException;
import java.io.OutputStream;
import java.util.Arrays;

import org.apache.xmlgraphics.image.GraphicsUtil;

/**
 * Helper class for encoding bitmap images.
 */
public class ImageEncodingHelper {

    private static final ColorModel DEFAULT_RGB_COLOR_MODEL = new ComponentColorModel(
            ColorSpace.getInstance(ColorSpace.CS_sRGB),
            false, false, ColorModel.OPAQUE, DataBuffer.TYPE_BYTE);

    private final RenderedImage image;
    private ColorModel encodedColorModel;
    private boolean firstTileDump;
    private boolean enableCMYK;
    private boolean isBGR;
    private boolean isKMYC;
    private boolean outputbw;
    private boolean bwinvert;

    /**
     * Main constructor
     * @param image the image
     */
    public ImageEncodingHelper(RenderedImage image) {
        this(image, true);
        outputbw = true;
    }

    /**
     * Main constructor
     * @param image the image
     * @param enableCMYK true to enable CMYK, false to disable
     */
    public ImageEncodingHelper(RenderedImage image, boolean enableCMYK) {
        this.image = image;
        this.enableCMYK = enableCMYK;
        determineEncodedColorModel();
    }

    /**
     * Returns the associated image.
     * @return the image
     */
    public RenderedImage getImage() {
        return this.image;
    }

    /**
     * Returns the native {@link ColorModel} used by the image.
     * @return the native color model
     */
    public ColorModel getNativeColorModel() {
        return getImage().getColorModel();
    }

    /**
     * Returns the effective {@link ColorModel} used to encode the image. If this is different
     * from the value returned by {@link #getNativeColorModel()} this means that the image
     * is converted in order to encode it because no native encoding is currently possible.
     * @return the effective color model
     */
    public ColorModel getEncodedColorModel() {
        return this.encodedColorModel;
    }

    /**
     * Indicates whether the image has an alpha channel.
     * @return true if the image has an alpha channel
     */
    public boolean hasAlpha() {
        return image.getColorModel().hasAlpha();
    }

    /**
     * Indicates whether the image is converted during encodation.
     * @return true if the image cannot be encoded in its native format
     */
    public boolean isConverted() {
        return getNativeColorModel() != getEncodedColorModel();
    }

    private void writeRGBTo(OutputStream out) throws IOException {
        boolean encoded = encodeRenderedImageWithDirectColorModelAsRGB(image, out);
        if (encoded) {
            return;
        }
        encodeRenderedImageAsRGB(image, out, outputbw, bwinvert);
    }

    public static void encodeRenderedImageAsRGB(RenderedImage image, OutputStream out)
            throws IOException {
        encodeRenderedImageAsRGB(image, out, false, false);
    }

    /**
     * Writes a RenderedImage to an OutputStream by converting it to RGB.
     * @param image the image
     * @param out the OutputStream to write the pixels to
     * @throws IOException if an I/O error occurs
     */
    public static void encodeRenderedImageAsRGB(RenderedImage image, OutputStream out,
            boolean outputbw, boolean bwinvert) throws IOException {
        Raster raster = getRaster(image);
        Object data;
        int nbands = raster.getNumBands();
        int dataType = raster.getDataBuffer().getDataType();
        switch (dataType) {
        case DataBuffer.TYPE_BYTE:
            data = new byte[nbands];
            break;
        case DataBuffer.TYPE_USHORT:
            data = null;
            break;
        case DataBuffer.TYPE_INT:
            data = new int[nbands];
            break;
        case DataBuffer.TYPE_FLOAT:
            data = new float[nbands];
            break;
        case DataBuffer.TYPE_DOUBLE:
            data = new double[nbands];
            break;
        default:
            throw new IllegalArgumentException("Unknown data buffer type: " + dataType);
        }

        ColorModel colorModel = image.getColorModel();
        int w = image.getWidth();
        int h = image.getHeight();
        int numDataElements = 3;
        if (colorModel.getPixelSize() == 1 && outputbw) {
            numDataElements = 1;
        }

        byte[] buf = new byte[w * numDataElements];

        for (int y = 0; y < h; y++) {
            int idx = -1;
            for (int x = 0; x < w; x++) {
                int rgb = colorModel.getRGB(raster.getDataElements(x, y, data));
                if (numDataElements > 1) {
                    buf[++idx] = (byte)(rgb >> 16);
                    buf[++idx] = (byte)(rgb >> 8);
                } else if (bwinvert && rgb == -1) {
                    rgb = 1;
                }
                buf[++idx] = (byte)(rgb);
            }
            out.write(buf);
        }
    }

    /**
     * Writes a RenderedImage to an OutputStream. This method optimizes the encoding
     * of the {@link DirectColorModel} as it is returned by {@link ColorModel#getRGBdefault}.
     * @param image the image
     * @param out the OutputStream to write the pixels to
     * @return true if this method encoded this image, false if the image is incompatible
     * @throws IOException if an I/O error occurs
     */
    public static boolean encodeRenderedImageWithDirectColorModelAsRGB(
            RenderedImage image, OutputStream out) throws IOException {
        ColorModel cm = image.getColorModel();
        if (cm.getColorSpace() != ColorSpace.getInstance(ColorSpace.CS_sRGB)) {
            return false; //Need to go through color management
        }
        if (!(cm instanceof DirectColorModel)) {
            return false; //Only DirectColorModel is supported here
        }
        DirectColorModel dcm = (DirectColorModel)cm;
        final int[] templateMasks = new int[]
                {0x00ff0000 /*R*/, 0x0000ff00 /*G*/, 0x000000ff /*B*/, 0xff000000 /*A*/};
        int[] masks = dcm.getMasks();
        if (!Arrays.equals(templateMasks, masks)) {
            return false; //no flexibility here right now, might never be used anyway
        }

        Raster raster = getRaster(image);
        int dataType = raster.getDataBuffer().getDataType();
        if (dataType != DataBuffer.TYPE_INT) {
            return false; //not supported
        }

        int w = image.getWidth();
        int h = image.getHeight();

        int[] data = new int[w];
        byte[] buf = new byte[w * 3];
        for (int y = 0; y < h; y++) {
            int idx = -1;
            raster.getDataElements(0, y, w, 1, data);
            for (int x = 0; x < w; x++) {
                int rgb = data[x];
                buf[++idx] = (byte)(rgb >> 16);
                buf[++idx] = (byte)(rgb >> 8);
                buf[++idx] = (byte)(rgb);
            }
            out.write(buf);
        }

        return true;
    }

    private static Raster getRaster(RenderedImage image) {
        if (image instanceof BufferedImage) {
            return ((BufferedImage)image).getRaster();
        } else {
            //Note: this copies the image data (double memory consumption)
            //TODO Investigate encoding in stripes: RenderedImage.copyData(WritableRaster)
            return image.getData();
        }
    }

    /**
     * Converts a byte array containing 24 bit RGB image data to a grayscale
     * image.
     *
     * @param raw
     *            the buffer containing the RGB image data
     * @param width
     *            the width of the image in pixels
     * @param height
     *            the height of the image in pixels
     * @param bitsPerPixel
     *            the number of bits to use per pixel
     * @param out the OutputStream to write the pixels to
     *
     * @throws IOException if an I/O error occurs
     */
    public static void encodeRGBAsGrayScale(
            byte[] raw, int width, int height, int bitsPerPixel, OutputStream out)
    throws IOException {
        int pixelsPerByte = 8 / bitsPerPixel;
        int bytewidth = (width / pixelsPerByte);
        if ((width % pixelsPerByte) != 0) {
            bytewidth++;
        }

        //TODO Rewrite to encode directly from a RenderedImage to avoid buffering the whole RGB
        //image in memory
        byte[] linedata = new byte[bytewidth];
        byte ib;
        for (int y = 0; y < height; y++) {
            ib = 0;
            int i = 3 * y * width;
            for (int x = 0; x < width; x++, i += 3) {

                // see http://www.jguru.com/faq/view.jsp?EID=221919
                double greyVal = 0.212671d * (raw[i] & 0xff) + 0.715160d
                        * (raw[i + 1] & 0xff) + 0.072169d
                        * (raw[i + 2] & 0xff);
                switch (bitsPerPixel) {
                case 1:
                    if (greyVal < 128) {
                        ib |= (byte) (1 << (7 - (x % 8)));
                    }
                    break;
                case 4:
                    greyVal /= 16;
                    ib |= (byte) ((byte) greyVal << ((1 - (x % 2)) * 4));
                    break;
                case 8:
                    ib = (byte) greyVal;
                    break;
                default:
                    throw new UnsupportedOperationException(
                            "Unsupported bits per pixel: " + bitsPerPixel);
                }

                if ((x % pixelsPerByte) == (pixelsPerByte - 1)
                        || ((x + 1) == width)) {
                    linedata[(x / pixelsPerByte)] = ib;
                    ib = 0;
                }
            }
            out.write(linedata);
        }
    }

    private boolean optimizedWriteTo(OutputStream out)
            throws IOException {
        if (this.firstTileDump) {
            Raster raster = image.getTile(0, 0);
            DataBuffer buffer = raster.getDataBuffer();
            if (buffer instanceof DataBufferByte) {
                byte[] bytes = ((DataBufferByte) buffer).getData();
                // see determineEncodingColorModel() to see why we permute B and R here
                if (isBGR) {
                    byte[] bytesPermutated = new byte[bytes.length];
                    for (int i = 0; i < bytes.length; i += 3) {
                        bytesPermutated[i] = bytes[i + 2];
                        bytesPermutated[i + 1] = bytes[i + 1];
                        bytesPermutated[i + 2] = bytes[i];
                    }
                    out.write(bytesPermutated);
                } else if (isKMYC) {
                    byte[] bytesPermutated = new byte[bytes.length];
                    for (int i = 0; i < bytes.length; i += 4) {
                        bytesPermutated[i] = bytes[i + 3];
                        bytesPermutated[i + 1] = bytes[i + 2];
                        bytesPermutated[i + 2] = bytes[i + 1];
                        bytesPermutated[i + 3] = bytes[i];
                    }
                    out.write(bytesPermutated);
                } else {
                    out.write(bytes);
                }
                return true;
            }
        }
        return false;
    }

    /**
     * Indicates whether the image consists of multiple tiles.
     * @return true if there are multiple tiles
     */
    protected boolean isMultiTile() {
        int tilesX = image.getNumXTiles();
        int tilesY = image.getNumYTiles();
        return (tilesX != 1 || tilesY != 1);
    }

    /**
     * Determines the color model used for encoding the image.
     */
    protected void determineEncodedColorModel() {
        this.firstTileDump = false;
        this.encodedColorModel = DEFAULT_RGB_COLOR_MODEL;

        ColorModel cm = image.getColorModel();
        ColorSpace cs = cm.getColorSpace();

        int numComponents = cm.getNumComponents();

        if (!isMultiTile()) {
            if (numComponents == 1 && cs.getType() == ColorSpace.TYPE_GRAY) {
                if (cm.getTransferType() == DataBuffer.TYPE_BYTE) {
                    this.firstTileDump = true;
                    this.encodedColorModel = cm;
                }
            } else if (cm instanceof IndexColorModel) {
                if (cm.getTransferType() == DataBuffer.TYPE_BYTE) {
                    this.firstTileDump = true;
                    this.encodedColorModel = cm;
                }
            } else if (cm instanceof ComponentColorModel
                    && (numComponents == 3 || (enableCMYK && numComponents == 4))
                    && !cm.hasAlpha()) {
                Raster raster = image.getTile(0, 0);
                DataBuffer buffer = raster.getDataBuffer();
                SampleModel sampleModel = raster.getSampleModel();
                if (sampleModel instanceof PixelInterleavedSampleModel) {
                    PixelInterleavedSampleModel piSampleModel;
                    piSampleModel = (PixelInterleavedSampleModel)sampleModel;
                    int[] offsets = piSampleModel.getBandOffsets();
                    for (int i = 0; i < offsets.length; i++) {
                        if (offsets[i] != i && offsets[i] != offsets.length - 1 - i) {
                            //Don't encode directly as samples are not next to each other
                            //i.e. offsets are not 012 (RGB) or 0123 (CMYK)
                            // let also pass 210 BGR and 3210 (KYMC); 3210 will be skipped below
                            // if 210 (BGR) the B and R bytes will be permuted later in optimizeWriteTo()
                            return;
                        }
                    }
                    // check if we are in a BGR case; this is added here as a workaround for bug fix
                    // http://bugs.sun.com/bugdatabase/view_bug.do?bug_id=6549882 that causes some PNG
                    // images to be loaded as BGR with the consequence that performance was being impacted
                    this.isBGR = false;
                    if (offsets.length == 3 && offsets[0] == 2 && offsets[1] == 1 && offsets[2] == 0) {
                        this.isBGR = true;
                    }
                    // make sure we did not get here due to a KMYC image
                    if (offsets.length == 4 && offsets[0] == 3 && offsets[1] == 2 && offsets[2] == 1
                            && offsets[3] == 0) {
                        isKMYC = true;
                    }
                }
                if (cm.getTransferType() == DataBuffer.TYPE_BYTE
                        && buffer.getOffset() == 0
                        && buffer.getNumBanks() == 1) {
                    this.firstTileDump = true;
                    this.encodedColorModel = cm;
                }
            }
        }

    }

    /**
     * Encodes the image and writes everything to the given OutputStream.
     * @param out the OutputStream
     * @throws IOException if an I/O error occurs
     */
    public void encode(OutputStream out) throws IOException {
        if (!isConverted()) {
            if (optimizedWriteTo(out)) {
                return;
            }
        }
        writeRGBTo(out);
    }

    /**
     * Encodes the image's alpha channel. If it doesn't have an alpha channel, an
     * {@link IllegalStateException} is thrown.
     * @param out the OutputStream
     * @throws IOException if an I/O error occurs
     */
    public void encodeAlpha(OutputStream out) throws IOException {
        if (!hasAlpha()) {
            throw new IllegalStateException("Image doesn't have an alpha channel");
        }
        Raster alpha = GraphicsUtil.getAlphaRaster(image);
        DataBuffer buffer = alpha.getDataBuffer();
        if (buffer instanceof DataBufferByte) {
            out.write(((DataBufferByte)buffer).getData());
        } else {
            throw new UnsupportedOperationException(
                    "Alpha raster not supported: " + buffer.getClass().getName());
        }
    }

    /**
     * Writes all pixels (color components only) of a RenderedImage to an OutputStream.
     * @param image the image to be encoded
     * @param out the OutputStream to write to
     * @throws IOException if an I/O error occurs
     */
    public static void encodePackedColorComponents(RenderedImage image, OutputStream out)
                throws IOException {
        ImageEncodingHelper helper = new ImageEncodingHelper(image);
        helper.encode(out);
    }

    /**
     * Create an ImageEncoder for the given RenderImage instance.
     * @param img the image
     * @return the requested ImageEncoder
     */
    public static ImageEncoder createRenderedImageEncoder(RenderedImage img) {
        return new RenderedImageEncoder(img);
    }

    /**
     * ImageEncoder implementation for RenderedImage instances.
     */
    private static class RenderedImageEncoder implements ImageEncoder {

        private final RenderedImage img;

        public RenderedImageEncoder(RenderedImage img) {
            this.img = img;
        }

        public void writeTo(OutputStream out) throws IOException {
            ImageEncodingHelper.encodePackedColorComponents(img, out);
        }

        public String getImplicitFilter() {
            return null; //No implicit filters with RenderedImage instances
        }
    }

    public void setBWInvert(boolean v) {
        bwinvert = v;
    }
}
